core/src/main/java/org/apache/iceberg/DataFiles.java

```diff
@@ -218,7 +219,9 @@ public Builder withFormat(FileFormat newFormat) {
     }
 
     public Builder withPartition(StructLike newPartition) {
-      this.partitionData = copyPartitionData(spec, newPartition, partitionData);
+      if (isPartitioned) {
+        this.partitionData = copyPartitionData(spec, newPartition, partitionData);
+      }
```

Comment:
Only copy if the spec is partitionable.

Reply:
I would rather we fix this in one of the other ways we discussed. This ends up feeling a little strange to me since you can call "withPartition" and the result is ignored. I don't like when functions can take arguments but then not doing anything with them.

Reply:
@aokolnychyi Do you mind taking a look at this change?

I have a chat with @RussellSpitzer offline, and would love to have your thoughts on this.

The current implementation
- works similar to how we do check for other methods inside the builder
    - https://github.com/apache/iceberg/blob/73ee251be87094183cf062f95d042c06b2acc6d5/core/src/main/java/org/apache/iceberg/DataFiles.java#L238-L245
    -  https://github.com/apache/iceberg/blob/73ee251be87094183cf062f95d042c06b2acc6d5/core/src/main/java/org/apache/iceberg/DataFiles.java#L259-L266
- Compatible with builder's whole building logics 
    - builder assign `partitionData` to be `null` if the spec is not partitioned: https://github.com/apache/iceberg/blob/73ee251be87094183cf062f95d042c06b2acc6d5/core/src/main/java/org/apache/iceberg/DataFiles.java#L141
    - And it do another check and pass in `null` to the final `GenericDataFile`: https://github.com/apache/iceberg/blob/73ee251be87094183cf062f95d042c06b2acc6d5/core/src/main/java/org/apache/iceberg/DataFiles.java#L294

Alternatively, we are thinking:
```java
GenericRecord emptyRecord = GenericRecord.create(spec.schema());
delegate = writerFactory.newDataWriter(outputFile, spec, emptyRecord);
```
at around here: https://github.com/apache/iceberg/blob/1e5abcece00d835235dccf0b902ffd988cbded0d/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java#L594